src/hotspot/share/classfile/classFileParser.cpp

```diff
@@ -4076,26 +4076,6 @@ void ClassFileParser::check_super_class_access(const InstanceKlass* this_klass,
       return;
     }
 
-    // If the loader is not the boot loader then throw an exception if its
-    // superclass is in package jdk.internal.reflect and its loader is not a
-    // special reflection class loader
-    if (!this_klass->class_loader_data()->is_the_null_class_loader_data()) {
-      PackageEntry* super_package = super->package();
-      if (super_package != nullptr &&
-          super_package->name()->fast_compare(vmSymbols::jdk_internal_reflect()) == 0 &&
-          !java_lang_ClassLoader::is_reflection_class_loader(this_klass->class_loader())) {
-        ResourceMark rm(THREAD);
-        Exceptions::fthrow(
-          THREAD_AND_LOCATION,
-          vmSymbols::java_lang_IllegalAccessError(),
-          "class %s loaded by %s cannot access jdk/internal/reflect superclass %s",
-          this_klass->external_name(),
-          this_klass->class_loader_data()->loader_name_and_id(),
-          super->external_name());
-        return;
-      }
-    }
-
```

Comment:
This code should not be removed. The spec for this code should now be:
```
  // If the loader is not the boot loader then throw an exception if its
  // superclass is in package jdk.internal.reflect 
```
All we need do is remove the check:
```
&& !java_lang_ClassLoader::is_reflection_class_loader
```

Reply:
Oh that's a good find.  Maybe we should write a test for this, if as I assume there isn't one already.

Reply:
> // If the loader is not the boot loader then throw an exception if its
> // superclass is in package jdk.internal.reflect

This will never happen.   `jdk.internal.reflect` is a non-exported package in `java.base` module.  If another module `M` defines a class whose superclass is in `jdk.internal.reflect` package, `java.base` must export `jdk.internal.reflect` package to `M` for access.   Otherwise, it will fail the super access check, as done in the check below this deleted code.

```
Reflection::VerifyClassAccessResults vca_result =
     Reflection::verify_class_access(this_klass, InstanceKlass::cast(super), false);
```

Reply:
`test/hotspot/jtreg/runtime/AccessCheckSuper.java` is one test that fails the super class access check.  @lfoltan may know whether there are more tests besides this one.

Reply:
This code was necessary back when this was the `sun.reflect` package and it was not able to be encapsulated (before modules) to prevent untrusted class loaders from leaking the `MagicAccessorImpl` class hierarchy.

Reply:
Okay ... so such subclass was and remains illegal and is now detected slightly later. That suggests to me this special case was in the wrong place and should have been inside `Reflection::verify_class_access`.

Anyway thanks for clarifying full removal is in fact what we want.